File: 2019/07 July/18th.py
# Good morning! Here's your coding interview problem for today.
#
# This problem was asked by Facebook.
#
# Given a list of integers, return the largest product that can be made by multiplying any three integers.
#
# For example, if the list is [-10, -10, 5, 2], we should return 500, since that's -10 * -10 * 5.
#
# You can assume the list has at least three integers.


def max_sum(l):
    l = sorted(l)
    out = [l[0] * l[1] * l[-1], l[-1] * l[-2] * l[-3]]  # More efficient ways but this is the simplest
    return max(out)


if __name__ == '__main__':
    print(max_sum([-10, -10, 5, 2]))
    print(max_sum([10, -10, 5, 2]))
    print(max_sum([-10, -10, -5, -2]))
    print(max_sum([1, 3, -5, -10]))

File: 2020/05 May/18th.go
//Good morning! Here's your coding interview problem for today.
//
//This problem was asked by Amazon.
//
//Given a sorted array arr of distinct integers, return the lowest index i for which arr[i] == i. Return null if there is no such index.
//
//For example, given the array [-5, -3, 2, 3], return 2 since arr[2] == 2. Even though arr[3] == 3, we return 2 since it's the lowest index.

package main

import "fmt"

func findSmallestIndex(arr [] int) int{
	for i, num := range arr{
		if i == num {
			return i
		}
	}
	return -1
}

func main() {
	slc  := [] int {-5, -3, 2, 3}
	fmt.Println(findSmallestIndex(slc))
	return
}